src/tir/ir/index_map.cc

```diff
@@ -147,6 +147,37 @@ IndexMap IndexMap::Inverse(Array<Range> initial_ranges) const {
   return inverse;
 }
 
+/*!
+ * \brief Evaluator to compute the mapped indices of a given index map.
+ */
+class IndexMapEvaluator : public DataTypeLegalizer {
```

Comment:
It inherits from `DataTypeLegalizer` that contains extra logic than `StmtMutator`, I think it's better to be separated from `IRSubstitute`. Perhaps we can introduce another utility function `SubstituteWithDataTypePromotion`

Reply:
Hmm, good point.  I had been thinking that the `DataTypeLegalizer` would have the necessary follow-up steps that should be performed after every substitution in order to have valid TIR, rather having extra logic as well.  Since there is extra logic, that would make sense to me to have the separate `SubstituteWithDataTypeLegalization`.